src/ethereum/dai.rs

```diff
@@ -46,47 +46,67 @@ static DAI_CONTRACT_ADDRESS_ROPSTEN: Lazy<Address> = Lazy::new(|| {
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Asset {
     pub amount: Amount,
-    pub contract_address: DaiContractAddress,
-    pub chain_id: ChainId,
+    pub chain: Chain,
 }
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub enum DaiContractAddress {
+pub enum Chain {
     Mainnet,
     Kovan,
     Rinkeby,
     Ropsten,
-    Local(Address),
+    Local {
+        chain_id: u32,
+        dai_contract_address: Address,
+    },
 }
 
-impl From<DaiContractAddress> for Address {
-    fn from(var: DaiContractAddress) -> Self {
-        match var {
-            DaiContractAddress::Mainnet => *DAI_CONTRACT_ADDRESS_MAINNET,
-            DaiContractAddress::Kovan => *DAI_CONTRACT_ADDRESS_KOVAN,
-            DaiContractAddress::Rinkeby => *DAI_CONTRACT_ADDRESS_RINKEBY,
-            DaiContractAddress::Ropsten => *DAI_CONTRACT_ADDRESS_ROPSTEN,
-            DaiContractAddress::Local(address) => address,
+impl Chain {
+    pub fn new(chain_id: ChainId, dai_contract_address: Address) -> Self {
```

Comment:
That's an interesting design. I think I like it.

Reply:
Yeah, I'm pretty happy with it. We could even go a little further and make this return an error if you pass a public chain ID with the wrong DAI contract address. But this seems fine anyway.